### ISSUE TITLE
Reading a Memory File before it is opened

### ISSUE THREAD

**Issue by MikeFair** [spazzarama/SharedMemory#6]
If a consumer app attempts to open a Shared Memory File before the producer has created it, a FileNotFound Exception is thrown.  A little digging didn't turn up any way to gracefully test/check for a MemoryMappedFile before trying to open it (or even better get some kind of file watcher or WaitEvent for it to show up).

So I used this function, but it feels like it belongs in the library somewhere:

        static Boolean TryOpenExisting(out SharedMemory.CircularBuffer buf, String name, TimeSpan ts)
        {
            System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
            if (ts.TotalMilliseconds > 0) { sw.Start();  }

            do
            {
                try
                {
                    buf = new SharedMemory.CircularBuffer(name);
                    return true;
                }
                catch (System.IO.FileNotFoundException e)
                {
                    buf = null;
                }
                catch (Exception e)
                {
                    throw e;
                }
            } while (sw.Elapsed < ts);

            return (buf != null);
        }

Here's how I used it in the consumer code:

                    if (consumer == null) {
                        TryOpenExisting(out consumer, name: "MySharedMemory", ts: TimeSpan.FromSeconds(5f));
                        if (consumer != null) { cbuf = new byte[consumer.NodeBufferSize]; }
                    }
                    if (consumer != null)
                    { 
                        while ((consumer.Read<byte>(cbuf) > 0)) {
                            ........
                        }
                    }

**Comment by spazzarama** [comment]
Good idea, I'll implement a wait event tho for it.

**Comment by MikeFair** [comment]
Would it make sense to also add an Event or callback for when data is written to the buffer as well?

I thinking through talking to multiple processes at once which I can't help but expect to require one file per writer.

The problem with the current implementation is I couldn't see how, if reading from multiple consumers do something other than a CPU intensive poll.  I suppose though, if there were multiple processes writing to the same file where there was just a single process reading it.  The library makes sure each of their writes gets onto the list, then the reader would simply read the messages off the list; and there'd be no need to select from multiple files...

If each process (thread?) had its own file... then I suppose multiple processes could all write to that processes individual file to send it a message...

**Comment by spazzarama** [comment]
@MikeFair Best to keep the conversation over in the other issue, but these are all ideas worth exploring. There are a number of uses for the base underlying class, we can always create variations that support the different scenarios, or else via configuration.

**Comment by MikeFair** [comment]
@spazzarama Good point; I'm deleting the comment to keep this and #7 separated (I've also got a better idea on how to go about handling #7 now). :)
